byterun/win32.c

```diff
@@ -84,17 +84,99 @@ static void caml_win32_sys_error(int errnum)
   caml_raise_sys_error(msg);
 }
 
+#define STATIC_BUFFER_SIZE 1024
+int win32_utf8_read(int fd, void * buf, int n) {
+  HANDLE hConsole = (HANDLE)_get_osfhandle(fd);
+  WCHAR szBuffer[STATIC_BUFFER_SIZE];
+  int real_n = (n > STATIC_BUFFER_SIZE << 2 ? STATIC_BUFFER_SIZE : n >> 2);
+  DWORD nread;
+  if (!ReadConsole(hConsole, szBuffer, real_n, &nread, NULL))
+    return -1;
+  return WideCharToMultiByte(CP_UTF8, 0, szBuffer, nread, buf, n, NULL, NULL);
+}
+
+int win32_utf8_write(int fd, void * buf, int n) {
+  HANDLE hConsole = (HANDLE)_get_osfhandle(fd);
+  int length;
+  WCHAR szBuffer[STATIC_BUFFER_SIZE];
+  LPWSTR buffer;
+  DWORD dwOffset = 0;
+  DWORD dwWritten;
+
+  length = MultiByteToWideChar(CP_UTF8, MB_ERR_INVALID_CHARS, buf, n, NULL, 0);
+  if (length == 0 || length == n) {
+    /* Either invalid UTF-8 or ASCII */
+    return write(fd, buf, n);
+  } else {
+    if (length > STATIC_BUFFER_SIZE) {
+      /* Just resort to write if for some weird reason we run out of memory */
+      if (!(buffer = (LPWSTR)malloc(length * sizeof(WCHAR))))
+        return write(fd, buf, n);
+    } else {
+      buffer = szBuffer;
+    }
+    /* This call must succeed */
+    MultiByteToWideChar(CP_UTF8, 0, buf, n, buffer, length);
+    while (length > 0) {
+      if (WriteConsole(hConsole, buffer + dwOffset, length, &dwWritten, NULL)) {
+        length -= dwWritten;
+        dwOffset += dwWritten;
+      } else {
+        /* Discarding errors at this point */
+        length = 0;
+      }
+    }
+    if (length > STATIC_BUFFER_SIZE)
```

Comment:
`length = 0` here!! 😄

Reply:
Oops :blush:

Reply:
It should be `buffer != szBuffer`... I'll push a change in a bit.